### PR TITLE
Add photon loss channel

### PR DESCRIPTION
**Pull request by Hugh-888**

### CODE REVIEW

File: src/deepquantum/photonic/operation.py [hunk 1]
```diff
@@ -49,6 +50,10 @@ def tensor_rep(self, x: torch.Tensor) -> torch.Tensor:
         """Get the tensor representation of the state."""
         return x.reshape([-1] + [self.cutoff] * self.nmode)
 
+    def matrix_rep(self, x: torch.Tensor) -> torch.Tensor:
+        """Get the density matrix representation of the state."""
+        return x.reshape([-1] + [self.cutoff] * 2 * self.nmode)
```
Comment (sansiro77) [inline src/deepquantum/photonic/operation.py]: Operation都加入den_mat，从而把密度矩阵的张量表示也合并到tensor_rep

File: src/deepquantum/photonic/operation.py [hunk 2]
```diff
@@ -132,14 +137,45 @@ def update_matrix_state(self) -> torch.Tensor:
     def op_state_tensor(self, x: torch.Tensor) -> torch.Tensor:
         """Perform a forward pass for state tensors."""
         nt = len(self.wires)
-        matrix = self.update_matrix_state().reshape(self.cutoff ** nt, self.cutoff ** nt)
+        matrix = self.update_matrix_state() # kraus matrix
+        if nt > 1:
+            matrix = matrix.reshape(self.cutoff ** nt, self.cutoff ** nt)
+        if len(x.size()) -1 == self.nmode: # fock tensor
```
Comment (sansiro77) [inline src/deepquantum/photonic/operation.py]: 也就可以用self.den_mat判断

File: src/deepquantum/photonic/gate.py [hunk 3]
```diff
@@ -1591,3 +1591,92 @@ def phi(self):
...
+        requires_grad: bool = False,
+        noise: bool = False,
+        mu: float = 0,
+        sigma: float = 0.1
+    ) -> None:
+        super().__init__(name='Loss', inputs=inputs, nmode=nmode, wires=wires, cutoff=cutoff,
+                         requires_grad=requires_grad, noise=noise, mu=mu, sigma=sigma)
+        self.npara = 1
+        self.init_para(inputs)
+        bs = BeamSplitter(inputs=[torch.arccos(torch.sqrt(self.t)), torch.pi], nmode=2, wires=None, cutoff=cutoff, requires_grad=requires_grad,
```
Comment (sansiro77) [inline src/deepquantum/photonic/gate.py]: torch.pi要改为与self.t一样device的Tensor，且BS的requires_grad永远为False

File: src/deepquantum/photonic/gate.py [hunk 4]
```diff
@@ -1591,3 +1591,92 @@ def phi(self):
...
+        elif not isinstance(inputs, (torch.Tensor, nn.Parameter)):
+            inputs = torch.tensor(inputs, dtype=torch.float)
+        if self.noise:
+            inputs = inputs + torch.normal(self.mu, self.sigma, size=(1, )).squeeze()
+        return inputs
+
+    def get_matrix(self, t: Any) -> torch.Tensor:
+        """Get the local unitary matrix acting on creation operators."""
+        t = self.inputs_to_tensor(t)
+        return t.reshape(1, 1)
```
Comment (sansiro77) [inline src/deepquantum/photonic/gate.py]: 应该是根号t？

File: src/deepquantum/photonic/gate.py [hunk 5]
```diff
@@ -1591,3 +1591,92 @@ def phi(self):
...
+
+    def update_matrix(self) -> torch.Tensor:
+        """Update the local unitary matrix acting on creation operators."""
+        matrix = self.get_matrix(self.t)
+        self.matrix = matrix.detach()
+        return matrix
+
+    def get_matrix_state(self, t: torch.Tensor) -> torch.Tensor:
+        """Get the local Kraus matrix acting on Fock state density matrix."""
+        inputs=[torch.arccos(torch.sqrt(t)), torch.pi]
```
Comment (sansiro77) [inline src/deepquantum/photonic/gate.py]: torch.pi也要改为Tensor

File: src/deepquantum/photonic/gate.py [hunk 6]
```diff
@@ -1591,3 +1591,92 @@ def phi(self):
...
+        return matrix_2
+
+    def init_para(self, inputs: Any = None) -> None:
+        """Initialize the parameters."""
+        noise = self.noise
+        self.noise = False
+        t = self.inputs_to_tensor(inputs)
+        self.noise = noise
+        if self.requires_grad:
+            self.theta = nn.Parameter(t)
```
Comment (sansiro77) [inline src/deepquantum/photonic/gate.py]: self.t

File: src/deepquantum/photonic/circuit.py [hunk 7]
```diff
@@ -97,6 +100,8 @@ def __init__(
         self.wires_homodyne = []
         self._draw_nstep = None
 
+        self.nlc = 0
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: 改用self._nmode_loss？

File: src/deepquantum/photonic/gate.py [hunk 8]
```diff
@@ -1591,3 +1592,118 @@ def phi(self):
...
+        else:
+            self.register_buffer('t', t)
+
+    def forward(
```
Comment (sansiro77) [inline src/deepquantum/photonic/gate.py]: 应该重载op_gaussian，而非forward。op_gaussian目前其实还有bug，不支持跨mode的门

File: src/deepquantum/photonic/circuit.py [hunk 9]
```diff
@@ -681,11 +700,23 @@ def get_unitary(self) -> torch.Tensor:
             operators = self._operators_tdm
         else:
             operators = self.operators
+        nlc = 0
         for op in operators:
-            if u is None:
-                u = op.get_unitary()
+            if isinstance(op, Loss):
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: 只修改op.gate的nmode和wires。nmode直接修改为self.nmode + self._nmode_loss，wires依然通过计数

File: src/deepquantum/photonic/circuit.py [hunk 10]
```diff
@@ -681,11 +700,23 @@ def get_unitary(self) -> torch.Tensor:
...
+                    u = op.gate.get_unitary()
+                else:
+                    u = torch.block_diag(u, torch.eye(1, dtype=u.dtype, device=u.device))
+                    u = op.gate.get_unitary() @ u
             else:
-                u = op.get_unitary() @ u
+                if u is None:
+                    u = op.get_unitary()
+                else:
+                    op.nmode = self.nmode + nlc
```
Comment (sansiro77) [inline src/deepquantum/photonic/circuit.py]: 不修改op.nmode，酉矩阵与self._nmode_loss的单位阵直和

**Review by sansiro77 — COMMENTED** [review]
`_get_all_fock_basis`之前加入了cutoff的限制，所以如果max_photon - nphotons有大于cutoff的情况，没法构造相应的final_state。可能需要对`_get_all_fock_basis`加入cutoff参数，为None时才用self.cutoff

File: src/deepquantum/photonic/gate.py (hunk 3 above)
Comment (sansiro77) [reply src/deepquantum/photonic/gate.py]: BeamSplitterTheta和BeamSplitterPhi也有类似bug

File: src/deepquantum/photonic/gate.py (hunk 6 above)
Comment (sansiro77) [reply src/deepquantum/photonic/gate.py]: 感觉Loss的变分参数还是用theta，不然t应该是个0-1的数，但没有在梯度更新时给相应的约束。给用户的接口可以留透射率和dB的表示

File: src/deepquantum/photonic/gate.py (hunk 4 above)
Comment (Hugh-888) [reply src/deepquantum/photonic/gate.py]: loss channel 的 get_matrix 和update_matrix 函数实际没起作用

File: src/deepquantum/photonic/circuit.py (hunk 9 above)
Comment (Jooyuza) [reply src/deepquantum/photonic/circuit.py]: 线路的unitary是一个逐渐变大的过程，每有一个loss才变大一维。这样设计对常规只在线路最后每模加loss的情况更友好

File: src/deepquantum/photonic/gate.py (hunk 4 above)
Comment (sansiro77) [reply src/deepquantum/photonic/gate.py]: 至少按照`get_matrix`是得到matrix acting on creation operators的定义，得是根号t，这样对QumodeCircuit做get_unitary可以得到一个形式上考虑光损失的非酉矩阵

File: src/deepquantum/photonic/circuit.py (hunk 9 above)
Comment (sansiro77) [reply src/deepquantum/photonic/circuit.py]: 截屏2024-11-22 15 26 18
block_diag比较慢，基本上50mode以内反而更慢

**Review by sansiro77 — APPROVED** [review]
Good job!
